units/units_util.hpp

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright (c) 2019-2021,
+Lawrence Livermore National Security, LLC;
+See the top-level NOTICE for additional details. All rights reserved.
+SPDX-License-Identifier: BSD-3-Clause
+*/
+#pragma once
+
+#include "units_decl.hpp"
+
+namespace units {
+namespace detail {
+
+    template<int Bits>
+    struct bitfield {
+        static constexpr int32_t min = -(static_cast<int32_t>(1U << Bits) / 2);
+        static constexpr int32_t max = static_cast<int32_t>(1U << Bits) / 2 - 1;
+        template<class T>
+        static constexpr bool plus_overflows(const T& a, const T& b)
+        {
+            return ((b > 0) && (a > max - b)) || ((b < 0) && (a < min - b));
+        }
+        template<class T>
+        static constexpr bool minus_overflows(const T& a, const T& b)
+        {
+            return ((b < 0) && (a > max + b)) || ((b > 0) && (a < min + b));
+        }
+        template<class T>
+        static constexpr bool times_overflows(const T& a, const T& b)
+        {
+            return ((a == -1) && (b == min)) || ((b == -1) && (a == min)) ||
+                (a > max / b) || ((a < min / b));
+        }
+    };
+
+    constexpr bool times_overflows(const unit_data& a, const unit_data& b)
+    {
+        return (
```

Comment:
wondering if it would be good to make a set of constexpr constants in units_decl.hpp that has the field sizes,  That might make the code for making the base type bigger simpler since you wouldn't need any preprocessor macros in the `unit_data` class and this code could avoid magic numbers.